File: algorithms/109. Convert Sorted List to Binary Search Tree/solution.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        int count = 0;
        ListNode * cur = head;
        while ( cur ) {
            ++count;
            cur = cur->next;
        }
        cur = head;
        return inorder(count, cur);
    }

    TreeNode * inorder(int count, ListNode * & cur) {
        if ( 0 == count ) return NULL;
        TreeNode * root = new TreeNode(0);
        root->left = inorder(count / 2, cur);
        root->val = cur->val;
        cur = cur->next;
        root->right = inorder(count - count / 2 - 1, cur);
        return root;
    }
};

File: algorithms/109. Convert Sorted List to Binary Search Tree/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sortedListToBST(self, head):
        """
        :type head: ListNode
        :rtype: TreeNode
        """
        def inorder(count):
            if count == 0:
                return
            root = TreeNode(0)
            root.left = inorder(count / 2)
            root.val = cur[0].val
            cur[0] = cur[0].next
            root.right = inorder(count - count / 2 - 1)
            return root

        cur = head
        count = 0
        while cur:
            count += 1
            cur = cur.next
        cur = [head]
        return inorder(count)

# 32 / 32 test cases passed.
# Status: Accepted
# Runtime: 266 ms
# beats 70.62 %
